File: leet-code/profitableSchemes.go
package main

import (
	"fmt"
	"math"
)

func profitableSchemes(G int, P int, group []int, profit []int) int {
	mod := int(math.Pow(10, 9) + 7)
	// 初始化一个三维表
	N := len(profit)
	dp := make([][][]int, N+1)
	for i := range dp {
		dp[i] = make([][]int, G+1)
		for j := range dp[i] {
			dp[i][j] = make([]int, P+1)
		}
	}

	// 初始化数据，利润为0的方案只有一种，就是什么罪都不犯
	for i := 0; i <= N; i++ {
		for j := 0; j <= G; j++ {
			dp[i][j][0] = 1
		}
	}
	fmt.Println(dp)
	// 填充三维表
	for i := 1; i <= N; i++ {
		for j := 1; j <= G; j++ {
			for k := 0; k <= P; k++ {
				// 如果不犯，方案不变
				dp[i][j][k] = dp[i-1][j][k]
				// 人数大于0才有意义
				if j-group[i-1] >= 0 {
					// 如果利润为负数，没有意义
					pw := max(k-profit[i-1], 0)
					// 状态转义方程，犯这个罪的方案
					dp[i][j][k] += dp[i-1][j-group[i-1]][pw]
				}
				dp[i][j][k] %= mod
			}
		}
	}
	return dp[N][G][P] % mod
}

// 压缩空间维度
func profitableSchemes2(G int, P int, group []int, profit []int) int {
	mod := int(math.Pow(10, 9) + 7)
	// 初始化一个三维表
	N := len(profit)
	dp := make([][]int, G+1)
	for i := range dp {
		dp[i] = make([]int, P+1)
	}

	// 初始化数据，利润为0的方案只有一种，就是什么罪都不犯
	for j := 0; j <= G; j++ {
		dp[j][0] = 1
	}

	fmt.Println(dp)
	// 填充三维表
	for i := 1; i <= N; i++ {
		for j := G; j >= 1; j-- {
			for k := P; k >= 0; k-- {
				// 人数大于0才有意义
				if j-group[i-1] >= 0 {
					// 如果利润为负数，没有意义
					pw := max(k-profit[i-1], 0)
					// 状态转义方程，犯这个罪的方案
					dp[j][k] += dp[j-group[i-1]][pw]
				}
				dp[j][k] %= mod
			}
		}
	}
	return dp[G][P] % mod
}

File: leet-code/profitableSchemes.js
/**
 * @param {number} G
 * @param {number} P
 * @param {number[]} group
 * @param {number[]} profit
 * @return {number}
 */
var profitableSchemes = function (G, P, group, profit) {
  const len = profit.length
  const mod = Math.pow(10, 9) + 7
  const dp = new Array(len + 1).fill(0).map(item => {
    return new Array(G + 1).fill(0).map(item => {
      return new Array(P + 1).fill(0)
    })
  })

  // 初始化，所有目标利润为0的都一种方案，就是啥都不干。。
  for (let i = 0; i <= len; i++) {
    for (let j = 0; j <= G; j++) {
      dp[i][j][0] = 1
    }
  }
  for (let i = 1; i <= len; i++) {
    for (let j = 1; j <= G; j++) {
      for (let k = 0; k <= P; k++) {
        // 这个是不选择i这个罪,方案数不会变
        dp[i][j][k] = dp[i - 1][j][k]
        if (j - group[i - 1] >= 0) {
          // 如果利润为负数，没有意义
          const nk = Math.max(k - profit[i - 1], 0)
          dp[i][j][k] += dp[i - 1][j - group[i - 1]][nk]
        }
        dp[i][j][k] = dp[i][j][k] % mod
      }
    }
  }

  console.log('dp==', dp, len, G, P)
  return dp[len][G][P] % mod
}

// console.log(profitableSchemes(5, 1, [2, 2], [2, 3]))

// 压缩空间维度
var profitableSchemes = function (G, P, group, profit) {
  const len = profit.length
  const mod = Math.pow(10, 9) + 7
  const dp = new Array(G + 1).fill(0).map(item => {
    return new Array(P + 1).fill(0)
  })

  // 初始化，所有目标利润为0的都一种方案，就是啥都不干。。
  for (let j = 0; j <= G; j++) {
    dp[j][0] = 1
  }

  for (let i = 1; i <= len; i++) {
    for (let j = G; j >= group[i - 1]; j--) {
      for (let k = P; k >= 0; k--) {
        // 如果利润为负数，没有意义
        const nk = Math.max(k - profit[i - 1], 0)
        dp[j][k] += dp[j - group[i - 1]][nk]
        dp[j][k] = dp[j][k] % mod
      }
    }
  }

  console.log('dp==', dp, len, G, P)
  return dp[G][P] % mod
}

console.log(profitableSchemes(5, 1, [2, 2], [2, 3]))
